server/parameters/ConfigFactory.java

```diff
@@ -0,0 +1,120 @@
+ *
+ */
+
+package com.vaticle.typedb.core.server.parameters;
+
+import com.vaticle.typedb.common.yaml.Yaml;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.server.parameters.util.Option;
+
+import java.io.FileNotFoundException;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CLI_OPTION_REQUIRES_VALUE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_FILE_NOT_FOUND;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_YAML_MUST_BE_MAP;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.ENV_VAR_NOT_FOUND;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+import static com.vaticle.typedb.core.server.common.Constants.CONFIG_PATH;
+
+public class ConfigFactory {
+
+    public static Config create(ConfigParser parser) {
+        return create(new HashSet<>(), parser);
+    }
+
+    public static Config create(Set<Option> overrides, ConfigParser parser) {
+        return create(CONFIG_PATH, overrides, parser);
+    }
+
+    public static Config create(Path file, Set<Option> overrides, ConfigParser parser) {
+        Yaml.Map yaml = merge(file, overrides);
+        substituteEnvVars(yaml);
+        return parser.parse(yaml, "");
+    }
+
+    private static Yaml.Map merge(Path file, Set<Option> overrides) {
+        Yaml.Map yaml = readFile(file);
+        Yaml.Map yamlOverrides = convertOverrides(overrides);
+        for (String key: yamlOverrides.keys()) {
+            set(yaml, key, yamlOverrides.get(key));
+        }
+        return yaml;
+    }
+
+    private static Yaml.Map readFile(Path file) {
+        try {
+            Yaml config = Yaml.load(file);
+            if (!config.isMap()) throw TypeDBException.of(CONFIG_YAML_MUST_BE_MAP);
+            return config.asMap();
+        } catch (FileNotFoundException e) {
+            throw TypeDBException.of(CONFIG_FILE_NOT_FOUND, file);
+        }
+    }
+
+    private static Yaml.Map convertOverrides(Set<Option> options) {
+        Set<String> keys = new HashSet<>();
+        for (Option option : options) {
+            if (!option.hasValue()) throw TypeDBException.of(CLI_OPTION_REQUIRES_VALUE, option);
+            keys.add(option.name());
+        }
+        Map<String, Yaml> yaml = new HashMap<>();
+        for (String key : keys) {
+            yaml.put(key, Yaml.load(get(options, key)));
+        }
+        return new Yaml.Map(yaml);
+    }
+
+    private static String get(Set<Option> source, String key) {
+        Set<String> values = iterate(source).filter(opt -> opt.name().equals(key))
+                .map(opt -> opt.stringValue().get()).toSet();
+        if (values.size() == 1) return values.iterator().next();
+        else return "[" + String.join(", ", values) + "]";
```

Comment:
oh cool i forgot we have support for merging multiple repeated cli args into a list in yaml!